client/file-manager/src/in_memory.rs

```diff
@@ -1,40 +1,167 @@
-use storage_hub_infra::types::{Chunk, FileProof, Key, Metadata};
+use std::collections::HashMap;
 
-use crate::traits::FileStorage;
+use common::types::HashT;
+use sp_core::H256;
+use storage_hub_infra::types::{Chunk, ChunkId, FileProof, Key, Leaf, Metadata};
 
-pub struct InMemoryFileStorage {}
+use sp_trie::{recorder::Recorder, MemoryDB, Trie, TrieDBBuilder, TrieLayout, TrieMut};
+use trie_db::TrieDBMutBuilder;
 
-impl InMemoryFileStorage {
+use crate::traits::{FileStorage, FileStorageError};
+
+pub struct FileData<T: TrieLayout + 'static> {
+    root: HashT<T>,
+    memdb: MemoryDB<T::Hash>,
+}
+
+impl<T: TrieLayout + 'static> FileData<T> {
+    pub fn get() {}
+}
+
+impl<T: TrieLayout + 'static> FileData<T> {
+    fn new() -> Self {
+        Self {
+            root: Default::default(),
+            memdb: MemoryDB::default(),
+        }
+    }
+
+    pub fn get_root(&self) -> H256 {
+        H256::from_slice(
+            self.root
+                .as_ref()
+                .try_into()
+                .expect("trie hash should be 32 bytes"),
+        )
+    }
+}
+
+pub struct InMemoryFileStorage<T: TrieLayout + 'static> {
+    pub metadata: HashMap<Key, Metadata>,
```

Comment:
These `HashMap`s will probably be accessed MANY times by concurrent threads. How about using [DashMap](https://github.com/xacrimon/dashmap)? It would also allow us to avoid using `RwLock` on the whole `InMemoryFileStorage`.

Reply:
Yeah, I've added a [task](https://www.notion.so/moonsonglabs/Use-DashMap-for-metadata-and-file_data-of-InMemoryFileStorage-940eac6e1211482599c85e6c0e3031bb?pvs=4) to our backlog.